charmhelpers/contrib/openstack/audits/openstack_security_guide.py

```diff
@@ -0,0 +1,216 @@
+    'nova': {
+        # Explicitly mentioned in security guide
+        '/etc/nova/nova.conf': {
+            'group': 'nova',
+            'mode': '640',
+        },
+        '/etc/nova/api-paste.ini': {
+            'group': 'nova',
+            'mode': '640',
+        },
+        '/etc/nova/rootwrap.conf': {
+            'group': 'nova',
+            'mode': '640',
+        },
+        # Other
+        '/etc/nova/nova-compute.conf': {
+            'group': 'nova',
+            'mode': '640',
+        },
+        '/etc/nova/logging.conf': {
+            'group': 'nova',
+            'mode': '640',
+        },
+        '/etc/nova/nm.conf': {
+            'group': 'nova',
+            'mode': '640',
+        },
+        '/etc/nova/*': {
+            'group': 'nova',
+            'mode': '640',
+        },
+    }
+}
+
+Ownership = collections.namedtuple('Ownership', 'owner group mode')
+
+
+@cached
+def _stat(file):
+    """
+    Get the Ownership information from a file.
+
+    :param file: The path to a file to stat
+    :type file: str
+    :returns: owner, group, and mode of the specified file
+    :rtype: Ownership
+    :raises subprocess.CalledProcessError: If the underlying stat fails
+    """
+    out = subprocess.check_output(['stat', '-c', '%U %G %a', file]).decode('utf-8')
+    return Ownership(*out.strip().split(' '))
+
+
+@cached
+def _config_ini(path):
+    """
+    Parse an ini file
+
+    :param path: The path to a file to parse
+    :type file: str
+    :returns: Configuration contained in path
+    :rtype: Dict
+    """
+    conf = configparser.ConfigParser()
+    conf.read(path)
+    return dict(conf)
+
+
+def _validate_file_ownership(owner, group, file_name):
+    """
+    Validate that a specified file is owned by `owner:group`.
+
+    :param owner: Name of the owner
+    :type owner: str
+    :param group: Name of the group
+    :type group: str
+    :param file_name: Path to the file to verify
+    :type file_name: str
+    """
+    try:
+        ownership = _stat(file_name)
+    except subprocess.CalledProcessError as e:
+        print("Error reading file: {}".format(e))
+        assert False, "Specified file does not exist: {}".format(file_name)
+    assert owner == ownership.owner, \
+        "{} has an incorrect owner: {} should be {}".format(
+            file_name, ownership.owner, owner)
+    assert group == ownership.group, \
+        "{} has an incorrect group: {} should be {}".format(
+            file_name, ownership.group, group)
+    print("Validate ownership of {}: PASS".format(file_name))
+
+
+def _validate_file_mode(mode, file_name):
+    """
+    Validate that a specified file has the specified permissions.
+
+    :param mode: file mode that is desires
+    :type owner: str
+    :param file_name: Path to the file to verify
+    :type file_name: str
+    """
+    try:
+        ownership = _stat(file_name)
+    except subprocess.CalledProcessError as e:
+        print("Error reading file: {}".format(e))
```

Comment:
print or log; not sure what the intention is here (i.e. if this is running on a unit, log is preferable so it turns up in the juju debug log?)

Reply:
Fun tidbit - STDOUT from an action goes into the unit's debug log :)

Reply:
nice.

Reply:
Another thing to discuss regarding print or log, would be for testing purposes - if we use log, then we can easily mock out the charmhelpers.log function as well, and validate pass/fail via that method if we wish.

Given that we can also just mock dependencies correctly and have the test assertions (audits) work normally, it doesn't seem like that strong of a reason.